flocker/node/agents/test/test_blockdevice_manager.py

```diff
@@ -157,3 +167,102 @@ def test_formatted_bad_type(self):
         blockdevice = self._get_free_blockdevice()
         with self.assertRaisesRegexp(MakeFilesystemError, blockdevice.path):
             self.manager_under_test.make_filesystem(blockdevice, 'myfakeyfs')
+
+    def test_bind_mount(self):
+        """
+        Files created in a bind mount are visible in the original folder and
+        vice versa.
+        """
+        src_directory = self._get_directory_for_mount()
+        target_directory = self._get_directory_for_mount()
+        self.manager_under_test.bind_mount(src_directory, target_directory)
+        for create, view in [(target_directory, src_directory),
+                             (src_directory, target_directory)]:
+            filename = str(uuid4())
+            new_file = create.child(filename)
+            new_file.touch()
+            self.expectThat(view.child(filename).exists(), Equals(True),
+                            'Created file not visible through bind mount.')
+
+    def test_failing_bind_mount(self):
+        """
+        Attempts to bind mount to a mountpoint that does not exist fail with a
+        ``BindMountError``.
+        """
+        src_directory = self._get_directory_for_mount()
+        target_directory = self._get_directory_for_mount().child('nonexistent')
+        with ExpectedException(BindMountError, '.*nonexistent.*'):
+            self.manager_under_test.bind_mount(src_directory, target_directory)
+
+    def test_remount(self):
+        """
+        Mounts remounted read-only cannot be written to until they are
+        remounted read-write.
+        """
+        blockdevice = self._get_free_blockdevice()
+        mountpoint = self._get_directory_for_mount()
+        self.manager_under_test.make_filesystem(blockdevice, 'ext4')
+        self.manager_under_test.mount(blockdevice, mountpoint)
+
+        first_file = mountpoint.child(str(uuid4()))
+        second_file = mountpoint.child(str(uuid4()))
+        third_file = mountpoint.child(str(uuid4()))
+
+        first_file.touch()
+        self.expectThat(first_file.exists(), Equals(True))
+
+        self.manager_under_test.remount(mountpoint, Permissions.READ_ONLY)
+        self.expectThat(first_file.exists(), Equals(True))
+
+        with ExpectedException(OSError):
+            second_file.touch()
+        self.expectThat(second_file.exists(), Equals(False))
+
+        self.manager_under_test.remount(mountpoint, Permissions.READ_WRITE)
+        self.expectThat(first_file.exists(), Equals(True))
+        self.expectThat(second_file.exists(), Equals(False))
+        third_file.touch()
+        self.expectThat(third_file.exists(), Equals(True))
```

Comment:
Interesting, I don't quite see what treeshape gets us here (just simpler than the manual .touch() commands?). Although there is definitely something we could do in this space.... not sure what it is.